### PR TITLE
Initial support for @storybook/server

### PR DESCRIPTION
**Pull request by jonspalmer**
Issue: Added a new "server" framework

## What I did

Added @storybook/server based on @storybook/html.
What's new is that the `configure` method takes a new fetchHtml argument that is responsible for fetching the html from the story.

## How to test

I will followup with a kitchen sink example. 

Usage is something like this:

```javascript
import { configure } from '@storybook/server';
import Stories from '../someStorySource';

const fetchHtml = async function(id, params) {
  const url = new URL('http://localhost:3000/storybook_preview_show');
  url.search = new URLSearchParams({id, ...params} ).toString();

  const response = await fetch(url);
  return  await response.text();
}

configure(fetchHtml, () => Stories, module);
```

## Open Questions/Issues

- My Typescript foo is weak. Could do with some 👀to make sure I got it right
- Tests? There don't seem to be tests for frameworks but maybe I'm missing something.
- Async Story loading? What's the right way to handle this? Probably needs first class support in core.

### CODE REVIEW

**Comment by shilman** [comment]
Amazing progress @bostanio! I'm guessing this will take some back and forth to get right, so I'll start things off:

1) Each framework has at least one example in `examples/` which provides an end-to-end test. This def needs one!

2) I don't think `configure` should take an extra argument. How about providing a separate function to configure it? (Yes, I get that making it required in configure helps you enforce that it's provided)

```
import { setFetchHtml } from '@storybook/server';
setFetchHtml(someFunction);
```

**Comment by bostanio** [comment]
@shilman Working on an example. Exploring the simplest 'app' to use as the backend. I'm guessing its express. 

Can certainly change the `configure` signature. I like forcing it in configure given its required but I can see the value in all frameworks having the same signature.

**Comment by shilman** [comment]
I'd also be open to a third `options` object argument to `configure`.

**Comment by bostanio** [comment]
> I'd also be open to a third `options` object argument to `configure`.

OK I'll go with that.

**Comment by bostanio** [comment]
Added a first pass at the example app. Would love some advice on how we might structure that example and how much of a 'kitchen sink' we want it to be.

File: app/server/src/client/preview/render.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  id,
+  selectedKind,
+  selectedStory,
+  showMain,
+  showError,
+  forceRender,
+}: RenderMainArgs) {
+  const params = storyFn();
+
+  const element = await fetchStoryHtml(id, params);
```
Comment (tmeasday) [inline app/server/src/client/preview/render.ts]: Should we show a useful error here if the user forgets to set this?

File: app/server/src/server/framework-preset-server.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    ...config,
+    module: {
+      ...config.module,
+      rules: [
+        ...config.module.rules,
+        {
+          test: /\.html$/,
+          use: [
+            {
+              loader: require.resolve('html-loader'),
```
Comment (tmeasday) [inline app/server/src/server/framework-preset-server.ts]: What do we need this for?

File: app/server/src/server/build.ts [hunk 3]
```diff
@@ -0,0 +1,4 @@
+import { buildStatic } from '@storybook/core/server';
+import options from './options';
+
+buildStatic(options);
```
Comment (tmeasday) [inline app/server/src/server/build.ts]: What do these files do? Were they copied over from `@storybook/html`?

**Review by tmeasday — CHANGES_REQUESTED** [review]
This is so cool!

File: app/server/src/client/preview/render.ts (hunk 1 above)
Comment (bostanio) [reply app/server/src/client/preview/render.ts]: Yup that's a great idea. It's entirely unrecoverable.

File: app/server/src/server/framework-preset-server.ts (hunk 2 above)
Comment (bostanio) [reply app/server/src/server/framework-preset-server.ts]: entirely copy pasta from @storybook/html. I was unclear for the need of them there. We can nuke right?

File: app/server/src/server/build.ts (hunk 3 above)
Comment (bostanio) [reply app/server/src/server/build.ts]: yup copied. I'm fuzzy on what this stuff is needed for. Should we nuke?

File: app/server/src/server/framework-preset-server.ts (hunk 2 above)
Comment (tmeasday) [reply app/server/src/server/framework-preset-server.ts]: I think so. This is to allow you to import `.html` files in storybook builds that use this addon. That's not something you would do with `@storybook/server`, right?

File: app/server/src/server/build.ts (hunk 3 above)
Comment (bostanio) [reply app/server/src/server/build.ts]: Do we need this at all?

File: app/server/src/client/preview/render.ts (hunk 1 above)
Comment (bostanio) [reply app/server/src/client/preview/render.ts]: Done

File: examples/server-kitchen-sink/client/storybook.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+const port = process.env.PORT || 8080;
+
+const fetchHtml = async (id, params) => {
+  const [component, story] = id.split('--').map(s => camelcase(s));
+
+  const url = new URL(`http://localhost:${port}/storybook_preview/${component}/${story}`);
+  url.search = new URLSearchParams(params).toString();
+
+  // eslint-disable-next-line no-undef
+  const response = await fetch(url);
```
Comment (bostanio) [inline examples/server-kitchen-sink/client/storybook.js]: What's the right way to fix this without the linting hack?

Comment (shilman) [reply examples/server-kitchen-sink/client/storybook.js]: https://github.com/eslint/eslint/issues/4015#issuecomment-334929169

File: app/server/src/server/build.ts (hunk 3 above)
Comment (shilman) [reply app/server/src/server/build.ts]: I think we need this:

```
  "bin": {
    "build-storybook": "./bin/build.js",
```

File: examples/server-kitchen-sink/client/storybook.js (hunk 4 above)
Comment (bostanio) [reply examples/server-kitchen-sink/client/storybook.js]: nice that was an easy fix

File: app/server/src/client/preview/index.ts [hunk 5]
```diff
@@ -33,10 +32,10 @@ const setRenderFecthAndConfigure: ClientApi['configure'] = (loader, module, opti
...
 
 export const configure: ClientApi['configure'] = setRenderFecthAndConfigure;
-export const addDecorator: ClientApi['addDecorator'] = api.clientApi.addDecorator;
-export const addParameters: ClientApi['addParameters'] = api.clientApi.addParameters;
-export const clearDecorators: ClientApi['clearDecorators'] = api.clientApi.clearDecorators;
-export const setAddon: ClientApi['setAddon'] = api.clientApi.setAddon;
-export const forceReRender: ClientApi['forceReRender'] = api.forceReRender;
-export const getStorybook: ClientApi['getStorybook'] = api.clientApi.getStorybook;
-export const raw: ClientApi['raw'] = api.clientApi.raw;
+export const { addDecorator } = api.clientApi;
```
Comment (shilman) [inline app/server/src/client/preview/index.ts]: combine into one?

```js
export const { addParameters, clearDecorators, ...etc } = api.clientApi;
```

Comment (bostanio) [reply app/server/src/client/preview/index.ts]: fixed

File: app/server/README.md [hunk 6]
```diff
@@ -0,0 +1,25 @@
+# Storybook for Server
```
Comment (tmeasday) [inline app/server/README.md]: I guess we might need to expand on this a bit

File: app/server/src/client/preview/index.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+const api = start(render);
+
+export const storiesOf: ClientApi['storiesOf'] = (kind, m) => {
+  return (api.clientApi.storiesOf(kind, m) as ReturnType<ClientApi['storiesOf']>).addParameters({
+    framework,
+  });
+};
+
+const setRenderFecthAndConfigure: ClientApi['configure'] = (loader, module, options) => {
```
Comment (tmeasday) [inline app/server/src/client/preview/index.ts]:
```suggestion
const setRenderFetchAndConfigure: ClientApi['configure'] = (loader, module, options) => {
```

File: app/server/src/client/preview/render.ts [hunk 8]
```diff
@@ -0,0 +1,61 @@
+import { document, fetch, Node } from 'global';
+import dedent from 'ts-dedent';
+import { RenderMainArgs, FetchStoryHtmlType } from './types';
+
+const rootElement = document.getElementById('root');
+
+let fetchStoryHtml: FetchStoryHtmlType = async (url, id, params) => {
```
Comment (tmeasday) [inline app/server/src/client/preview/render.ts]:
```suggestion
let fetchStoryHtml: FetchStoryHtmlType = async (url, path, params) => {
```

Let's call this path and allow it to be empty (not sure if we should drop the trailing slash in that case?

File: app/server/src/client/preview/render.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+  const fetchId = storyId || id;
+  const fetchParams = { ...params, ...storyParams };
+  const element = await fetchStoryHtml(url, fetchId, fetchParams);
+
+  showMain();
+  if (typeof element === 'string') {
+    rootElement.innerHTML = element;
+  } else if (element instanceof Node) {
+    // Don't re-mount the element if it didn't change and neither did the story
+    if (rootElement.firstChild === element && forceRender === true) {
```
Comment (tmeasday) [inline app/server/src/client/preview/render.ts]:
```suggestion
    if (rootElement.firstChild === element && forceRender !== true) {
```

Is this what you meant?

File: app/server/src/lib/compiler/decorators/knobs.ts [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+  value: any;
+  [x: string]: any;
+}
+
+function stringifyKnob(knob: StoryKnob) {
+  const { param, type, name, value, ...opts } = knob;
+  const knobParam = param || name; // Todo: can we do away with this?
+  const level = 2;
+  const stringifiedValue = stringifyObject(value, level);
+  // TODO: Add group
```
Comment (tmeasday) [inline app/server/src/lib/compiler/decorators/knobs.ts]: Do we need to fix this TODO?

File: app/server/src/server/framework-preset-server.ts [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+          test: /\.stories\.json$/,
+          use: [
+            {
+              loader: path.resolve(__dirname, './loader.js'),
+            },
+          ],
+        },
+      ],
+    },
+  };
```
Comment (tmeasday) [inline app/server/src/server/framework-preset-server.ts]: So is the idea that you add the `x.stories.json` files to your storybook directory? What about the idea of fetching the json on startup?

**Review by tmeasday — COMMENTED** [review]
Looks great; I have some questions around the loading of story JSON.

File: app/server/src/server/framework-preset-server.ts (hunk 11 above)
Comment (tmeasday) [reply app/server/src/server/framework-preset-server.ts]: Discussed this and the answer is it is better to let the external process (i.e. Rails) watch the relevant files and regenerate the JSON file, triggering webpack to rebuild when the stories change.

File: app/server/README.md (hunk 6 above)
Comment (tmeasday) [reply app/server/README.md]: Will open a separate ticket for this

**Comment by tmeasday** [comment]
@jonspalmer there are still two unresolved comments on that PR that we need to sort out:

1. Is this a mistake or did I misunderstand: https://github.com/storybookjs/storybook/pull/9250#discussion_r366776949

2. Do we need to deal with this TODO: https://github.com/storybookjs/storybook/pull/9250/files/be4d06c0cd69b2ed4a79ad390852770bfd6ccfeb#diff-83aae226cfb73b52d12dfd37ae15a411?

Also a change to `mdx.tsx` got dropped because `generateHrefWithHash` doesn't exist any more. I'm not sure if that's going to be a problem but I'll have a proper play with this once the above two questions are resolved.

**Comment by ndelangen** [comment]
@shilman I assume you got a ok from @tmeasday to merge? Looks like he still had a few concerns?

**Comment by shilman** [comment]
@ndelangen yeah he filed some follow-up issues

**Comment by tmeasday** [comment]
@ndelangen this was a bit confusing but @jonspalmer answered my questions on the other PR

**Pull request merged**